src/State/ForKeys.lua

```diff
@@ -0,0 +1,231 @@
+
+	- they were not previously present
+	- a dependency used during generation of this key has changed
+
+	It will recalculate those keys, storing information about any dependencies used
+    in the processor callback during value generation, and saving the new key to the
+    output array with the same value. If it is overwriting an older value, that older
+    value will be passed to the destructor for cleanup.
+
+	Finally, this function will find keys that are no longer present, and remove
+	their output key from the output table and pass them to the destructor.
+]]
+function class:update(): boolean
+	local inputIsState = self._inputIsState
+	local oldInputKeys = self._oldInputTable
+	local newInputKeys = self._inputTable
+	local keyOIMap = self._keyOIMap
+	local outputKeys = self._outputTable
+	local meta = self._meta
+
+	if inputIsState then
+		newInputKeys = newInputKeys:get(false)
+	end
+
+	local didChange = false
+
+	-- clean out main dependency set
+	for dependency in pairs(self.dependencySet) do
+		dependency.dependentSet[self] = nil
+	end
+	self._oldDependencySet, self.dependencySet = self.dependencySet, self._oldDependencySet
+	table.clear(self.dependencySet)
+
+	-- if the input table is a state object, add as dependency
+	if inputIsState then
+		self._inputTable.dependentSet[self] = true
+		self.dependencySet[self._inputTable] = true
+	end
+
+	-- STEP 1: find keys that were not previously present
+	for newInKey, _value in pairs(newInputKeys) do
+		-- get or create key data
+		local keyData = self._keyData[newInKey]
+		if keyData == nil then
+			keyData = {
+				-- we don't need strong references here - the main set does that
+				-- for us, so let's not introduce unnecessary leak opportunities
+				dependencySet = setmetatable({}, WEAK_KEYS_METATABLE),
+				oldDependencySet = setmetatable({}, WEAK_KEYS_METATABLE),
+				dependencyValues = setmetatable({}, WEAK_KEYS_METATABLE),
+			}
+			self._keyData[newInKey] = keyData
+		end
+
+		-- if an input key's previous value is non-nil, then there's no need to recalculate it
+		-- this also allows it to have a non-truthy value, which is important since we don't care
+		-- about the value; if we have dependencies then we also need to check if they've changed,
+		-- since if they have then the key has "changed".
+		local shouldRecalculate = oldInputKeys[newInKey] == nil
+
+		if not shouldRecalculate then
+			-- check if dependencies have changed
+			for dependency, oldValue in pairs(keyData.dependencyValues) do
+				-- if the dependency changed value, then this needs recalculating
+				if oldValue ~= dependency:get(false) then
+					shouldRecalculate = true
+					break
+				end
+			end
+		end
+
+		-- if we should recalculate the output by this point, do that
+		if shouldRecalculate then
+			keyData.oldDependencySet, keyData.dependencySet = keyData.dependencySet, keyData.oldDependencySet
+			table.clear(keyData.dependencySet)
+
+			local processOK, newOutKey, newMetaValue = captureDependencies(
```

Comment:
Should meta be allowed to be variadic? The current implementation only allows a single meta value. This seems restrictive for no good reason.